### PR TITLE
Add dockerfile and stuff

### PR DESCRIPTION
**Pull request by r0mainK**
So, in this PR there are 3 commits.

**Commit 1:**

Add Dockerfile and requirements file. I always like to install requirements separately from the library when building Docker images, hence the requirements. As you can see there is a certain amount of ENV variables, going to get to that. 

**Commit 2:**

Here I add the env variable logic. This is directly inspired from modelforge, the idea is that is way easier to centralize everything in one directory, which allows us to remove a lot of the logic by simply specifying environment variables. 

If you agree, I would like to go further, assume that Docker will be used in all cases, and remove all mentions of input/output paths in the command, keeping only:

- For the output of `preprocess`, replace `output-dir` with `dataset-name`
- For the input of `create_bow` thus we can just specify `dataset-name`, and `bow-name` for the output
- For the input of `train_hdp` or other model training just specify `bow-name` and optionally `experiment-name` 
- For the input of postprocessing commands specify  `bow-name` and optionnaly `experiemnt-name`, and that's it.

I think it would be better this way, and the data will be stored in a logical way:
```
/data
/data/features
/data/features/dataset_1.pkl
/data/features/dataset_2.pkl
/data/bow
/data/bow/bow_1/
/data/bow/bow_1/docs.bow_tm.txt
/data/bow/bow_1/docword.bow_tm.txt
/data/bow/bow_1/vocab.bow_tm.txt
/data/topics
/data/topics/bow_1/
/data/topics/bow_1/experiment_1/
...
```

**Commit 3:**

Updated the README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,58 @@
...
+
+You should also create a directory for all the data. In the following, we call this `/path/to/data`, and will be using defaults values provided by the container for inputs/outputs.
+
+### Preprocess
+
+Launch Babelfish, install all drivers and check you can see them:
+
+```
+docker run -d --rm --name tmexp_bblfshd --privileged -p 9432:9432 -v /var/lib/bblfshd:/var/lib/bblfshd bblfsh/bblfshd:latest-drivers
+docker exec -it tmexp_bblfshd bblfshctl driver install --all -f
```
Comment (m09) [inline README.md]: We don't need to install drivers in the `bblfsh/bblfshd:latest-drivers` image. It's also important that we don't have to because we only emit a docker restart when we restart it from the main container (otherwise we'd also have to install drivers there  and it would be much slower).

File: README.md [hunk 2]
```diff
@@ -1 +1,58 @@
...
+
+Start by cloning the repository, then building the docker image:
+
+```
+git clone https://github.com/src-d/tm-experiments
+```
+
+You should also create a directory for all the data. In the following, we call this `/path/to/data`, and will be using defaults values provided by the container for inputs/outputs.
+
+### Preprocess
```
Comment (m09) [inline README.md]: We should rely on either our `Makefile` or a `docker-compose` config file instead of having to launch several commands by hand. This is less error-prone (and less tiring while working :relaxed:)

File: tmexp/__main__.py [hunk 3]
```diff
@@ -136,13 +136,7 @@ def get_parser() -> argparse.ArgumentParser:
...
     )
     preprocess_parser.add_argument(
-        "--bblfsh-container",
-        help="Name of the Babelfish docker container.",
-        type=str,
-        default="tmexp_bblfshd",
-    )
-    preprocess_parser.add_argument(
-        "--bblfsh-host", help="Babelfish hostname.", type=str, default="0.0.0.0"
+        "--bblfsh-host", help="Babelfish hostname.", type=str, default="tmexp_bblfshd"
```
Comment (m09) [inline tmexp/__main__.py]: Since we want to fully dockerize, we should nuke this and retrieve the BBLFSH_HOSTNAME env variable that we'd set in Dockerfile/Makefile/docker-compose.yml

File: tmexp/__main__.py [hunk 4]
```diff
@@ -209,16 +203,23 @@ def get_parser() -> argparse.ArgumentParser:
...
-        required=True,
+        " distributions, defaults to TOPIC_DIR.",
+        default=None,
+        type=str,
+    )
+    train_hdp_parser.add_argument(
+        "--exp-name",
+        help="Name of the experiment, defaults to 'experiment_i' where i is the number "
+        "of other experiments for this dataset.",
+        default=None,
```
Comment (m09) [inline tmexp/__main__.py]: We can use the date (MM-DD-HH:MM) instead of a number as default (or as default prefix, which I like more): more informative, sorts with string sort, etc.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  && rm -rf /var/lib/apt/lists/*
+
+RUN curl -sSL https://get.docker.com/ | sh 
+
+RUN apt-get update \
+  && apt-get install -y python3-pip python3-dev \
+  && cd /usr/local/bin \
+  && ln -s /usr/bin/python3 python \
+  && apt-get remove -y .*-doc .*-man >/dev/null \
+  && apt-get autoremove -y \
```
Comment (m09) [inline Dockerfile]:
```suggestion
  && apt-get autoremove --purge -y \ 
```

File: tmexp/utils.py [hunk 6]
```diff
@@ -49,11 +55,20 @@ def create_logger(log_level: str, name: str) -> Logger:
     return logger
 
 
-def check_exists(file_path: str) -> None:
+def check_file_exists(file_path: str) -> None:
     if not os.path.exists(file_path):
         raise RuntimeError("File '%s' does not exists, aborting." % file_path)
```
Comment (m09) [inline tmexp/utils.py]: Not from this PR but we should use FileNotFoundError

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  && apt-get remove -y .*-doc .*-man >/dev/null \
+  && apt-get autoremove -y \ 
+  && apt-get clean \
+  && rm -rf /var/lib/apt/lists/*
+
+
+COPY requirements.txt package/
+COPY setup.py package/
+COPY README.md package/
+COPY tmexp package/tmexp/
```
Comment (m09) [inline Dockerfile:35]: This should probably come after the dependency install: it will be more volatile that the deps and prevent their caching

File: tmexp/preprocess.py [hunk 8]
```diff
@@ -221,7 +228,7 @@ def feature_extractor(uast_obj: Any) -> Iterator[Tuple[str, str]]:
                 if time.time() - start > bblfsh_timeout - 0.1 and attempt == 0:
                     logger.warn("Babelfish timed out, restarting the container ...")
                     subprocess.call(
-                        ["docker", "restart", bblfsh_container],
+                        ["docker", "restart", bblfsh_host],
                         stdout=open(subprocess.DEVNULL, "wb"),
```
Comment (m09) [inline tmexp/preprocess.py]: Did not notice last time but we don't need open with subprocess.DEVNULL.

File: README.md [hunk 9]
```diff
@@ -1 +1,58 @@
...
+Clone one or multiple repository in a directory, then launch gitbase:
+
+```
+docker run -d --rm --name gitbase -p 3306:3306 --link tmexp_bblfshd:tmexp_bblfshd -e BBLFSH_ENDPOINT=tmexp_bblfshd:9432 -v /path/to/repos:/opt/repos srcd/gitbase:latest
+```
+
+Finally, launch the preprocessing (we give the docker socket in order to be able to relaunch Babelfish, however this is a temporary hack):
+
+```
+docker run --rm -i -v /var/run/docker.sock:/var/run/docker.sock -v /path/to/data:/data --link tmexp_bblfshd:tmexp_bblfshd --link gitbase:gitbase tmexp preprocess -r repo
```
Comment (m09) [inline README.md]: Now that we have 3 containers and 3 links, it's probably better to switch to the non-deprecated network instead of link, but since this is already working I'm ok with leaving it as is. Just mentioning it for a future experiment :)

**Review by m09 — CHANGES_REQUESTED** [review]
It's a great idea to dockerize indeed, the switch to the ML Cluster will be way faster :+1: 

Regarding env variables and arguments: let's dockerize completely and fix the output dirs also completely (no env variables for them). Right now it's verbose and a bit hard to follow (defaults in the dockerfile, Makefile, .utils, CLI args). We won't ever use anything else so let's nuke the complexity :)

File: README.md (hunk 1 above)
Comment (r0mainK) [reply README.md]: Even the first time we use the image ?

Comment (m09) [reply README.md]: Yup, it's stateless :)

Comment (r0mainK) [reply README.md]: Ah oke, my bad will remove :)

File: tmexp/__main__.py (hunk 3 above)
Comment (r0mainK) [reply tmexp/__main__.py]: Yes agreed

Comment (r0mainK) [reply tmexp/__main__.py]: Agreed, we should also do the same for Gitbase no ?

File: tmexp/__main__.py (hunk 4 above)
Comment (r0mainK) [reply tmexp/__main__.py]: done

File: Dockerfile (hunk 5 above)
Comment (r0mainK) [reply Dockerfile]: done

File: tmexp/utils.py (hunk 6 above)
Comment (r0mainK) [reply tmexp/utils.py]: got it

File: tmexp/__main__.py (hunk 3 above)
Comment (m09) [reply tmexp/__main__.py]: Yes, missed it during review!

File: tmexp/preprocess.py (hunk 8 above)
Comment (r0mainK) [reply tmexp/preprocess.py]: ah we can just put `stdout=subprocess.DEVNULL` ?

File: README.md (hunk 9 above)
Comment (r0mainK) [reply README.md]: I've never used it, how does it work ?

File: tmexp/preprocess.py (hunk 8 above)
Comment (m09) [reply tmexp/preprocess.py]: I do think so

File: README.md (hunk 9 above)
Comment (m09) [reply README.md]: Basically you create a network and specify it for each container. For example, in `style-analyzer`, I used:

    docker network create --driver bridge report_0.1.1
    docker run --network report_0.1.1 -d --rm --name report_0.1.1_bblfshd --privileged bblfsh/bblfshd\:v2.11.0 --log-level DEBUG
    docker exec report_0.1.1_bblfshd bblfshctl driver install javascript docker://bblfsh/javascript-driver:v1.2.0
    docker run --network report_0.1.1 -e LOOKOUT_BBLFSHD=ipv4://report_0.1.1_bblfshd:9432 --name report_0.1.1 --entrypoint /bin/bash -it --rm -v /storage/hugo:/storage -v /home/hugo:/home ubuntu:18.04

The advantage is that all containers see each other directly in the network, no need for linking. It's a bit easier to maintain and it's not deprecated (link is). As I said since this works it's not necessary to tweak it right now, I was mentioning it to let you know for the next time :)

File: Dockerfile (hunk 7 above)
Comment (r0mainK) [reply Dockerfile:35]: oke

File: README.md (hunk 2 above)
Comment (r0mainK) [reply README.md]: True - although it is the only one which requires > 1 command, I agree it would be more convenient to use docker-compose/Makefile for this one

File: README.md (hunk 9 above)
Comment (r0mainK) [reply README.md]: Oke, no it looks cool - and idk that link as depreciated. Will probably include this in future PR, it makes sense to use it

File: README.md (hunk 1 above)
Comment (m09) [reply README.md]: BTW I didn't notice it during the review but since it's stateless we should not mount /var/lib/bblfshd

Comment (r0mainK) [reply README.md]: oh my bad, removing :+1:

File: README.md (hunk 9 above)
Comment (m09) [reply README.md]: I should mention that it has a “disadvantage”: it persists after all containers have been removed if it is not removed also.

Comment (r0mainK) [reply README.md]: ah, hm I guess it's not so bad but yeah got to keep it in mind then

File: README.md [hunk 10]
```diff
@@ -1 +1,70 @@
...
+```
+
+For all commands we only specify the required arguments, check the optional ones with `docker run --rm -i tmexp $CMD --help`
+
+### Preprocess
+
+This command will allow you to create a dataset from a cloned git repository. Before launching the command, you will thus need to clone one (or multiple) repository in a directory, as well as start the Babelfish and Gitbase servers:
+
+```
+make start RESOLVED_REPOS=~/path/to/cloned/repos
```
Comment (m09) [inline README.md]: `REPOS=~/path/to/cloned/repos make start` is safer (would work with a relative path)

File: README.md [hunk 11]
```diff
@@ -1 +1,70 @@
...
+```
+
+You can now launch the preprocessing withe the follow command _(we give the docker socket in order to be able to relaunch Babelfish, however this is a **temporary** hack)_:
```
Comment (m09) [inline README.md]:
```suggestion
You can now launch the preprocessing with the following command _(we give the docker socket in order to be able to relaunch Babelfish, however this is a **temporary** hack)_:
```

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,38 @@
+FROM ubuntu:18.04
+
+RUN mkdir /data
+
+ENV BBLFSH_HOSTNAME=tmexp_bblfshd \
+  BBLFSH_PORT=9432 \
+  GITBASE_HOSTNAME=gitbase \
```
Comment (m09) [inline Dockerfile]: We should probably prefix it by default like bblfshd to avoid conflict if we run another gitbase:
```suggestion
  GITBASE_HOSTNAME=tmexp_gitbase \
```
This requires some harmonization in docs and maybe some other places though. I'm fine if we do it at some other point, it's not that important.

File: tmexp/__main__.py [hunk 13]
```diff
@@ -47,9 +47,30 @@ def add_force_arg(cmd_parser: argparse.ArgumentParser) -> None:
 def add_dataset_arg(cmd_parser: argparse.ArgumentParser) -> None:
     cmd_parser.add_argument(
         "--dataset-name",
-        help="Name of the dataset, used as the second level input/output directory.",
+        help="Name of the dataset created by `preprocess`, defaults"
+        " to 'MM-DD-HH:MM-dataset'.",
+        default=None,
```
Comment (m09) [inline tmexp/__main__.py]:
```suggestion
        default=check_create_default("dataset"),
```
with the mentioned edits to this function later in the review.

File: tmexp/__main__.py [hunk 14]
```diff
@@ -47,9 +47,30 @@ def add_force_arg(cmd_parser: argparse.ArgumentParser) -> None:
 def add_dataset_arg(cmd_parser: argparse.ArgumentParser) -> None:
     cmd_parser.add_argument(
         "--dataset-name",
-        help="Name of the dataset, used as the second level input/output directory.",
+        help="Name of the dataset created by `preprocess`, defaults"
+        " to 'MM-DD-HH:MM-dataset'.",
```
Comment (m09) [inline tmexp/__main__.py]:
```suggestion
        " to '%(default)s'.",
```

File: tmexp/__main__.py [hunk 15]
```diff
@@ -47,9 +47,30 @@ def add_force_arg(cmd_parser: argparse.ArgumentParser) -> None:
...
+        default=None,
+        type=str,
```
Comment (m09) [inline tmexp/__main__.py]: Default type is str, no need to specify it.

File: tmexp/__main__.py [hunk 16]
```diff
@@ -47,9 +47,30 @@ def add_force_arg(cmd_parser: argparse.ArgumentParser) -> None:
...
+    )
+
+
+def add_bow_arg(cmd_parser: argparse.ArgumentParser) -> None:
```
Comment (m09) [inline tmexp/__main__.py]: Ditto

File: tmexp/__main__.py [hunk 17]
```diff
@@ -47,9 +47,30 @@ def add_force_arg(cmd_parser: argparse.ArgumentParser) -> None:
...
+    cmd_parser.add_argument(
+        "--bow-name",
+        help="Name of the BoW created by `create_bow`, defaults"
+        " to 'MM-DD-HH:MM-bow'.",
+        default=None,
+        type=str,
+    )
+
+
+def add_experiment_arg(cmd_parser: argparse.ArgumentParser) -> None:
```
Comment (m09) [inline tmexp/__main__.py]: Ditto

File: tmexp/create_bow.py [hunk 18]
```diff
@@ -30,9 +32,8 @@ def check_fraction(fraction: float, arg_name: str) -> None:
 
 
 def create_bow(
-    input_path: str,
-    output_dir: str,
-    dataset_name: str,
+    dataset_name: Optional[str],
```
Comment (m09) [inline tmexp/create_bow.py]: Type hint should be `str` since it's not allowed to be `None`.

File: tmexp/create_bow.py [hunk 19]
```diff
@@ -30,9 +32,8 @@ def check_fraction(fraction: float, arg_name: str) -> None:
...
+    bow_name: Optional[str],
```
Comment (m09) [inline tmexp/create_bow.py]: should also be `str`, the logic should be handled at the CLI level so that illegal `None` are not allowed.

File: tmexp/preprocess.py [hunk 20]
```diff
@@ -185,7 +192,7 @@ def feature_extractor(uast_obj: Any) -> Iterator[Tuple[str, str]]:
     if stem:
         stemmer = PorterStemmer()
     blacklisted_files: Set[str] = set()
-    client = bblfsh.BblfshClient("%s:%d" % (bblfsh_host, bblfsh_port))
+    client = bblfsh.BblfshClient("%s:%s" % (bblfsh_host, bblfsh_port))
```
Comment (m09) [inline tmexp/preprocess.py]: We cast the gitbase port to int but not the bblfsh port. We should streamline that just to avoid type confusion.

File: tmexp/utils.py [hunk 21]
```diff
@@ -49,9 +63,22 @@ def create_logger(log_level: str, name: str) -> Logger:
     return logger
 
 
-def check_exists(file_path: str) -> None:
+def check_create_default(name: Optional[str], out_type: str, logger: Logger) -> str:
```
Comment (m09) [inline tmexp/utils.py]: There is a race condition (although since we use only minutes it's probably fine to bet on it not happening in 10M runs): the name can change between two calls. We can store the name in a private module variable and only return new names based on it once we've initialized it on the first call. Also, if we apply the suggestion to call this directly in the parser, we can remove the two last args.

**Review by m09 — CHANGES_REQUESTED** [review]
Thanks for the quick address of the many points in the previous review :100: 

There are a few suggestions in the review. I didn't include all occurrences of something I think we should fix: the parser should not allow `None`s when it's not desired. It probably means that we should add some `required=True` depending on which parser receives which arg. We can then clarify type hints. As a rule of thumb, it's good for a function to document the legality of arguments in the type hints if possible (since Python's type system is mostly terribad it's not always the case).

Another small nitpick: it's also good not to include default=None and type=str since those are default values in the add_argument calls.

Those should be easy fixes, the core of the dockerization seems solid now, good job :+1:

File: README.md (hunk 10 above)
Comment (r0mainK) [reply README.md]: agreed, I actually created a `/path/to/data` when testing the code x)

File: Dockerfile (hunk 12 above)
Comment (r0mainK) [reply Dockerfile]: no it's fine, I think it is only here, the Makefile and the docs

File: tmexp/__main__.py (hunk 15 above)
Comment (r0mainK) [reply tmexp/__main__.py]: oke will remove `type=str` everywhere then

File: tmexp/preprocess.py (hunk 20 above)
Comment (r0mainK) [reply tmexp/preprocess.py]: oke

File: tmexp/__main__.py (hunk 13 above)
Comment (r0mainK) [reply tmexp/__main__.py]: done here and elsewhere

File: tmexp/utils.py (hunk 21 above)
Comment (r0mainK) [reply tmexp/utils.py]: done, added a CUR_TIME variable in utils

**Comment by r0mainK** [comment]
I think it should be good now, unless you see something else we could optimize ?

File: tmexp/create_bow.py [hunk 22]
```diff
@@ -44,15 +44,20 @@ def create_bow(
 ) -> None:
     logger = create_logger(log_level, __name__)
 
-    check_exists(input_path)
-    output_dir = os.path.join(output_dir, dataset_name)
+    if dataset_name is None:
```
Comment (m09) [inline tmexp/create_bow.py]: Sorry I was not clear on this point: we have 2 different cases for the `dataset_name` arg: one can have a default (`preprocess`), the other cannot and should be required (`create_bow`, since the name should already exist from existing output). It means we should add an argument to `add_dataset_arg` to know if we have `required=True` or if we have the current `default=…`. If we have those two different parser definitions, we don't need to check for `None` here and we're also certain that the user does not use a default by mistake.

**Review by m09 — CHANGES_REQUESTED** [review]
Almost good to go! One argument left to fix.

File: tmexp/create_bow.py (hunk 22 above)
Comment (r0mainK) [reply tmexp/create_bow.py]: ah yes my bad, I forgot that part, updating and merging

**Review by m09 — APPROVED** [review]

**Comment by r0mainK** [comment]
Oke done, merging when Travis is ready
